src/main/resources/include/hail/NativeObj.h

```diff
@@ -0,0 +1,52 @@
+#ifndef HAIL_NATIVEOBJ_H
+#define HAIL_NATIVEOBJ_H 1
+
+#include <memory>
+
+// Declare a function to be exported from a DLL with "C" linkage
+
+#define NATIVEMETHOD(cppReturnType, scalaClass, scalaMethod) \
+  extern "C" __attribute__((visibility("default"))) \
+    cppReturnType Java_is_hail_nativecode_##scalaClass##_##scalaMethod
+
+namespace hail {
+
+// The lifetime of off-heap objects may be controlled by any combination
+// of C++ std::shared_ptr<T> and Jvm-side NativePtr objects.  But classes
+// used in this way must inherit from NativeObj.
+
+class NativeObj :
+  public std::enable_shared_from_this<NativeObj> {
+  public:
+    // Objects managed with a Jvm-side NativePtr need to be destroyed
+    // without knowing their precise type, so they must have a virtual
+    // destructor.
+    virtual ~NativeObj() { }
+
+    // Subclasses should override getClassName, which may be helpful
+    // for debugging.
+    virtual const char* get_class_name() { return "NativeObj"; }
+    
+    // If a subclass wants to make some fields visible to Jvm Unsafe
+    // access, it can publish their offsets.  This is probably only
+    // useful for integer types, so we pass in one parameter to
+    // identify the fieldSize as 1, 2, 4, or 8 bytes.
+    virtual long get_field_offset(int fieldSize, const char* fieldName) { return(-1); }
+};
+
+// On the Jvm side, we don't have distinct classes/types for NativePtr
+// to different off-heap objects - we treat them all as NativeObjPtr.
+typedef std::shared_ptr<NativeObj> NativeObjPtr;
+
+#define MAKE_NATIVE(ObjT, ...) \
+  (std::static_pointer_cast<NativeObj>(std::make_shared<ObjT>(__VA_ARGS__)))
+
+// Upcast conversion of pointer-to-subclass to pointer-to-base
+template<typename T>
+inline NativeObjPtr toNativeObjPtr(const std::shared_ptr<T>& a) {
```

Comment:
No, because std::shared_ptr<T> is just a template class, not something built into the
language's type system, there aren't any implicit conversions.  That works for Base* and SubClass*, but not for shared_ptr<Base> and shared_ptr<SubClass>

The STL provides static_pointer_cast and dynamic_pointer_cast to do those conversions
explicitly.

Reply:
But the constructor
```C++
template< class Y >
shared_ptr( const shared_ptr<Y>& r ) noexcept;
```
((9) in http://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) is not explicit, so defines an implicit conversion from `shared_ptr<Y>` to `shared_ptr<T>` whenever `Y*` is implicitly convertible to `T*`. I tested this out with
```C++
#include <memory>

struct A {
  int i = 0;
};

struct B : A {};

int main(int argc, char *argv[])
{
  auto p = std::make_shared<B>();
  std::shared_ptr<A> p2 = p;
}
```
(Sorry if I'm being nitpicky, I'm just trying to understand C++ better.)

Reply:
You're right, that does work.  I guess that really eliminates the need for MAKE_NATIVE.

You do have to be a bit careful with shared_ptr to make sure that you don't go through
some sequence of conversions which goes through a naked T*, since that could lead to
having two control blocks (and two ref-counts) for a single object.  But I just tested that
and the implicit constructor from const shared_ptr<Y>& seems to do it right.

Reply:
I think before they were standardized, they had implicit conversions from smart pointers to raw pointers, and they got rid of them because they caused all kinds of problems such as the double control block (and `.get()` is pretty painless).